File: LeetCode/0904.Fruit-Into-Baskets.cpp
// 904. Fruit Into Baskets

class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int l=0, r=0, ans=0, n=fruits.size();
        unordered_map<int,int> mp;

        for(r=0; r<n; r++) {
            mp[fruits[r]]++;
            while(mp.size() > 2) {
                mp[fruits[l]]--;
                if(mp[fruits[l]] == 0) mp.erase(fruits[l]);
                l++;
            }
            ans = max(ans, r-l+1);
        }
        return ans;
    }
};

File: LeetCode/0904-Fruit-Into-Baskets.java
// 904. Fruit Into Baskets

class Solution {
    public int totalFruit(int[] fruits) {
        int n = fruits.length;
        int l = 0, r = 0, ans = 0;
        Map<Integer,Integer> mp = new HashMap<>();
        while (r < n) {
            int fR = fruits[r];
            mp.put(fR, mp.getOrDefault(fR, 0) + 1);
            while (mp.size() > 2) {
                int fL = fruits[l];
                mp.put(fL, mp.get(fL) - 1);
                if (mp.get(fL) == 0) mp.remove(fL);
                l++;
            }
            ans = Math.max(ans, r - l + 1);
            r++;
        }
        return ans;
    }
}
